### PR TITLE
User-defined generic category support (core + lua).

### PR DESCRIPTION
**Pull request by pirogronian**

### CODE REVIEW

**Comment by 375gnu** [comment]
Why to add additional ancestor to all object? Why not just add a categories as a private members to them?

**Comment by pirogronian** [comment]
Well... I always thought that good programmer should minimize code duplication.
What means "categories as a private members"?

**Comment by 375gnu** [comment]
You have `class Foo : Object`, instead of this I suggested `class Foo { private: Object* o; }`

But now think that your approach is better as you need some public functions to work with categories. Maybe you just need a better name like `UserCategoryHelper` instead of Object.

**Comment by pirogronian** [comment]
> Maybe you just need a better name like UserCategoryHelper instead of Object.
Maybe. However, I also tend to minimize names length :-) Moreover, I've already thought a lot about to unify these three classess, because it looks really similar in many ways.

**Comment by pirogronian** [comment]
These methods works, but for full comfort I want to fix one FIXME ;-)

File: src/celengine/object.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <string>
+#include <set>
+
+class UserCategory;
+
+class Object
+{
+    typedef std::set<UserCategory*> cat_set;
```
Comment (375gnu) [inline src/celengine/object.h]: According to Celestia's style guide it should be CategorySet.

File: src/celengine/object.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    cat_set *m_cats;
```
Comment (375gnu) [inline src/celengine/object.h]: I'd make it not a pointer but just `CategorySet m_cats`. Of course if you have tests that prove that it's worth to have such optimization here, then I'm okay with pointer.

File: src/celengine/object.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    bool _removeFromCategory(UserCategory*);
+public:
+    bool addToCategory(UserCategory*);
+    bool removeFromCategory(UserCategory*);
+    bool inCategory(UserCategory*) const;
+
+    bool addToCategory(const std::string&);
+    bool removeFromCategory(const std::string&);
+    bool inCategory(const std::string&) const;
+    int inCategories() const { return m_cats == nullptr ? 0 : m_cats->size(); }
```
Comment (375gnu) [inline src/celengine/object.h]: Misleading method name

File: src/celestia/celx_category.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include "celx_internal.h"
+#include "celx_category.h" 
+#include <celengine/category.h>
+
+static int category_tostring(lua_State *l)
+{
+    CelxLua celx(l);
+
+    UserCategory *c = *celx.getThis<UserCategory*>();
+    std::string s = fmt::format("[UserCategory:{0}]", c->name());
```
Comment (375gnu) [inline src/celestia/celx_category.cpp]: Could you replace it with `fmt::printf`? Just for consistence with other code. If `fmt::format` has advantages over `fmt::printf` then we'll allow its usage, but currently let's stick with a single API.

File: src/celestia/celx_internal.h [hunk 5]
```diff
@@ -94,7 +95,160 @@ class CelxLua
     CelxLua(lua_State* l);
     ~CelxLua() = default;
 
+    static int localIndex(int n) { return lua_upvalueindex(n); }
+
+    /**** push-on-stack methods ****/
+
+    int push()
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: It looks like these method should return `void`.

File: src/celestia/celx_internal.h [hunk 6]
```diff
@@ -94,7 +95,160 @@ class CelxLua
...
+    }
+    template <typename T> T *newUserDataArray(int n)
+    {
+        return reinterpret_cast<T*>(lua_newuserdata(m_lua, sizeof(T) * n));
+    }
+    template <typename T> T *newUserDataArray(T *a, int n)
+    {
+        T *p = reinterpret_cast<T*>(lua_newuserdata(m_lua, sizeof(T) * n));
+        for (int i = 0; i < n; i++)
+            p[i] = a[i];
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: Either `memcpy`  or `std::copy` are better choice.

File: src/celestia/celx_internal.h [hunk 7]
```diff
@@ -94,7 +95,160 @@ class CelxLua
...
+        int n = a.size();
+        int i = 0;
+        T *array = celx.newUserDataArray<T>(n);
+        for (auto &it : a)
+        {
+            array[i] = it.second;
+            i++;
+        }
+        celx.push((n - 1));
+        auto lambda = [](lua_State *l)
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: The function is too big to be lambda.

File: src/celestia/celx_internal.h [hunk 8]
```diff
@@ -94,7 +95,160 @@ class CelxLua
...
+    bool isInteger(int n = 0) const { return lua_isinteger(m_lua, n); }
+    bool isNumber(int n = 0) const { return lua_isnumber(m_lua, n); }
+    bool isBoolean(int n = 0) const { return lua_isboolean(m_lua, n); }
+    bool isString(int n = 0) const { return lua_isstring(m_lua, n); }
+    bool isUserData(int n = 0) const { return lua_isuserdata(m_lua, n); }
+    bool isValid(int) const;
+
+    /**** get methods ****/
+
+    int getInt(int n = 0) const { return lua_tointeger(m_lua, n); }
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: We already have `Celx::getSafe` family of funtions.

File: src/celestia/celx_internal.h [hunk 9]
```diff
@@ -149,8 +308,41 @@ class CelxLua
                         const char* errorMessage = "Boolean argument expected",
                         bool defaultValue = false);
 
-    LuaState* getLuaStateObject();
+    template<typename T> T *safeGetUserData(int index = 0,
+                                            FatalErrors errors = AllErrors,
+                                            const char *errorMessage = "User data expected")
+    {
+        if (!safeIsValid(index))  return nullptr;
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: Please don't use such style. There are of places in your code where it can be misunderstood due to this style.

File: src/celestia/celx_internal.h [hunk 10]
```diff
@@ -149,8 +308,41 @@ class CelxLua
...
+        if (!isUserData(index))
+        {
+            if (errors & WrongType)
+                doError(errorMessage);
+            return nullptr;
+        }
+        return reinterpret_cast<T*>(lua_touserdata(m_lua, index));
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: There are places with wrong style. Better:
```
if (isUserData(index))
    return reinterpret_cast<T*>(lua_touserdata(m_lua, index));

if (errors & WrongType)
    doError(errorMessage);

return nullptr;
```
Such code is much easier to understand.

File: src/celengine/object.h (hunk 2 above)
Comment (pirogronian) [reply src/celengine/object.h]: I just realized, that this class will be instantiated by thausend or more times, so its size should be minimized. However, I didn't test how much memory take an empty set. This could depend on compiler or library implementation.

File: src/celestia/celx_internal.h (hunk 5 above)
Comment (pirogronian) [reply src/celestia/celx_internal.h]: I did it for shorter code. Every push(...) function returns number of pushed elements (I know, for now it is 1 in every case), so I can write
`return push(...);`
instead of
`push();
return 1;`

File: src/celestia/celx_internal.h (hunk 8 above)
Comment (pirogronian) [reply src/celestia/celx_internal.h]: So, here are unsafe (but faster) functions... Perhaps I wrote them for symmetry.

File: src/celestia/celx_internal.h (hunk 9 above)
Comment (pirogronian) [reply src/celestia/celx_internal.h]: Do you mean `if - then` in one line?

Comment (375gnu) [reply src/celestia/celx_internal.h]: Yes

File: src/celestia/celx_internal.h (hunk 5 above)
Comment (375gnu) [reply src/celestia/celx_internal.h]: so more proper to write as return push(...)

File: src/celestia/celx_internal.h [hunk 11]
```diff
@@ -94,7 +95,137 @@ class CelxLua
...
+    {
+        CelxLua celx(m_lua);
+        int n = a.size();
+        T *array = celx.newUserDataArray<T>(n);
+        for (auto &it : a)
+        {
+            *array = value(it);
+            array++;
+        }
+        celx.push((n - 1));
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: extra braces

File: src/celengine/object.cpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+{
+    UserCategory *c = UserCategory::find(s);
+    if (c == nullptr)
+        return false;
+    return removeFromCategory(c);
+}
+
+bool Object::isInCategory(UserCategory *c) const
+{
+    if (m_cats == nullptr)  return false;
```
Comment (375gnu) [inline src/celengine/object.cpp]: wrong style

File: src/celengine/object.cpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    return m_cats->count(c);
```
Comment (375gnu) [inline src/celengine/object.cpp]: != 0 missed

Comment (pirogronian) [reply src/celengine/object.cpp]: Rather new line before `return false;`...

Comment (375gnu) [reply src/celengine/object.cpp]: I mean `m_cats->count(c) > 0`

**Comment by pirogronian** [comment]
category_getcategories is for now unimplemented due to Object <-> Selection incompatibilities.

File: src/celestia/celx_internal.h [hunk 14]
```diff
@@ -149,8 +291,42 @@ class CelxLua
...
+        if (!isType(i, celxClassId(*a)))
+        {
+            if (fatalErrors & WrongType)
+                doError(msg);
+            return nullptr;
+        }
+        return a;
+    }
 
+    template<typename T> T *getThis(FatalErrors fatalErrors = AllErrors,
```
Comment (375gnu) [inline src/celestia/celx_internal.h]: Please replace `getThis` with more meaningful name.

File: src/celengine/object.cpp [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+{
+    if (m_cats == nullptr)
+        m_cats = new CategorySet;
+    m_cats->insert(c);
+    return true;
+}
+
+bool Object::addToCategory(UserCategory *c)
+{
+    if (!_addToCategory(c))  return false;
```
Comment (375gnu) [inline src/celengine/object.cpp]: still wrong style

File: src/celengine/object.cpp [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        delete m_cats;
+        m_cats = nullptr;
+    }
+    return true;
+}
+
+bool Object::removeFromCategory(UserCategory *c)
+{
+    if (!_removeFromCategory(c))  return false;
```
Comment (375gnu) [inline src/celengine/object.cpp]: and here

File: src/celestia/celx_category.cpp [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    const char *n = c->name().c_str();
+    return celx.push(n);
+}
+
+static int category_createchild(lua_State *l)
+{
+    CelxLua celx(l);
+
+    UserCategory *c = *celx.getThis<UserCategory*>();
+    const char *n = celx.safeGetString(2);
```
Comment (375gnu) [inline src/celestia/celx_category.cpp]: check for nullptr

File: src/celestia/celx_object.cpp [hunk 18]
```diff
@@ -1408,21 +1409,69 @@ static int object_setatmosphere(lua_State* l)
...
+    Selection *s = celx.getThis<Selection>();
+    bool ret;
+    if (celx.isUserData(2))
+    {
+        UserCategory *c = *celx.getUserData<UserCategory*>(2);
+        ret = s->object()->addToCategory(c);
+    }
+    else
+    {
+        const char *n = celx.safeGetString(2, AllErrors, "Argument to object:addtocategory() must be string or userdata");
```
Comment (375gnu) [inline src/celestia/celx_object.cpp]: nullptr

File: src/celestia/celx_object.cpp [hunk 19]
```diff
@@ -1408,21 +1409,69 @@ static int object_setatmosphere(lua_State* l)
...
+    Selection *s = celx.getThis<Selection>();
+    bool ret;
+    if (celx.isUserData(2))
+    {
+        UserCategory *c = *celx.getUserData<UserCategory*>(2);
+        ret = s->object()->removeFromCategory(c);
+    }
+    else
+    {
+        const char *n = celx.safeGetString(2, AllErrors, "Argument to object:addtocategory() must be string or userdata");
```
Comment (375gnu) [inline src/celestia/celx_object.cpp]: nullptr

File: src/celestia/celx_category.cpp (hunk 17 above)
Comment (pirogronian) [reply src/celestia/celx_category.cpp]: I added additional check inside Celx_safeGetString.

File: src/celestia/celx_internal.h (hunk 14 above)
Comment (pirogronian) [reply src/celestia/celx_internal.h]: Hmm, I have no idea. Legacy <classname>_this was more meaningful?

File: src/celestia/celx_category.cpp (hunk 17 above)
Comment (375gnu) [reply src/celestia/celx_category.cpp]: I did not understood it, I'll create a new issue for tis case. I just need to make some checks before this.

File: src/celestia/celx_internal.h (hunk 14 above)
Comment (375gnu) [reply src/celestia/celx_internal.h]: Nope :)

Comment (375gnu) [reply src/celestia/celx_internal.h]: But I  don't have any good proposals so if you either don't have any, let's keep getThis.

File: src/celengine/category.cpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    for (auto it : m_catlist)
+        deleteCategory(it);
+}
+
+UserCategory::CategoryMap UserCategory::m_allcats;
+UserCategory::CategorySet UserCategory::m_roots;
+
+UserCategory *UserCategory::newCategory(const std::string &s, UserCategory *p)
+{
+    if (m_allcats.count(s))  return nullptr;
```
Comment (375gnu) [inline src/celengine/category.cpp]: count() > 0 and new line

File: src/celengine/category.cpp [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+UserCategory *UserCategory::createRoot(const std::string &n)
+{
+    return newCategory(n, nullptr);
+}
+
+UserCategory *UserCategory::find(const std::string &s)
+{
+    if (!m_allcats.count(s)) return nullptr;
```
Comment (375gnu) [inline src/celengine/category.cpp]: count > 0 and  new line

File: src/celengine/category.cpp [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+//    std::cout << "UserCategory::find(" << s << "): exists\n";
+    return (*m_allcats.find(s)).second;
```
Comment (375gnu) [inline src/celengine/category.cpp]: m_allcats.find(s)->second

File: src/celengine/category.cpp [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+bool UserCategory::deleteCategory(UserCategory *c)
+{
+    if (!find(c->name()))  return false;
```
Comment (375gnu) [inline src/celengine/category.cpp]: new line

File: src/celengine/category.h [hunk 24]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <set>
+#include <map>
+#include "object.h"
+
+class UserCategory
+{
+public:
+    typedef std::set<Object*>    ObjectSet;
```
Comment (375gnu) [inline src/celengine/category.h]: extra spaces

File: src/celengine/category.h [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    typedef std::set<UserCategory*> CategorySet;
+    typedef std::map<const std::string, UserCategory*> CategoryMap;
+
+private:
+    std::string   m_name;
+    ObjectSet      m_objlist;
+    CategorySet      m_catlist;
```
Comment (375gnu) [inline src/celengine/category.h]: extra spaces

File: src/celengine/category.h [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+
+    bool addObject(Object*);
+    bool removeObject(Object*);
+    const ObjectSet &objects() const { return m_objlist; }
+    const CategorySet &children() const { return m_catlist; }
+
+private:
+    bool _addObject(Object*);
+    bool _removeObject(Object*);
+    void _insertChild(UserCategory *);
```
Comment (375gnu) [inline src/celengine/category.h]: extra space

File: src/celengine/object.h [hunk 27]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <string>
+#include <set>
+
+class UserCategory;
+
+class Object
```
Comment (375gnu) [inline src/celengine/object.h]: As I said previously `Object` is too generic and doesn't reflect its purpose.

File: src/celestia/celx_internal.h [hunk 28]
```diff
@@ -149,8 +291,42 @@ class CelxLua
...
+                                         const char *msg = "Celx class expected")
+    {
+        T *a = safeGetUserData<T>(i, fatalErrors, msg);
+        if (!isType(i, celxClassId(*a)))
+        {
+            if (fatalErrors & WrongType)
+                doError(msg);
+            return nullptr;
+        }
+        return a;
```
Comment (375gnu) [inline src/celestia/celx_internal.h]:
```
    if (isType(i, celxClassId(*a)))
        return a;

    if (fatalErrors & WrongType)
        doError(msg);
    return nullptr;
```

Comment (375gnu) [reply src/celestia/celx_internal.h]: cause first we care about returned value

File: src/celestia/celx.cpp [hunk 29]
```diff
@@ -1233,13 +1235,15 @@ const char* Celx_SafeGetString(lua_State* l,
...
+    const char *s;
+    if (!lua_isstring(l, index) || 
+        ((s = lua_tostring(l, index)) == nullptr))
     {
         if (fatalErrors & WrongType)
             Celx_DoError(l, errorMsg);
         return nullptr;
     }
-    return lua_tostring(l, index);
+    return s;
```
Comment (375gnu) [inline src/celestia/celx.cpp]: please revert this change. if the object is `lua_isstring` then `lua_tostring` won't return nullptr. And you still return nullptr to the caller.

let me show a simple example:
```
#include <iostream>
#include <string>

const char* fn() {return nullptr;}

int main()
{
  std::string s = fn();
  std::cout << s << '\n';
}
```

Now compile and run it: `g++ t.cc && ./a.out`

You get:
```
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted
```

And that's why I'm telling you that Celx_SafeGetString is not so safe and you should check it's return.

File: src/celestia/celx_category.cpp (hunk 17 above)
Comment (375gnu) [reply src/celestia/celx_category.cpp]:
```
category:createchild()
-- or
category:createchild(1)
```
and celestia is aborted

because you did not check for nullptr

**Comment by pirogronian** [comment]
Understood.

**Comment by pirogronian** [comment]
This was really long day... However, I know there is at least one bug.

File: src/celengine/selection.h [hunk 30]
```diff
@@ -87,4 +96,9 @@ inline bool operator!=(const Selection& s0, const Selection& s1)
     return s0.type != s1.type || s0.obj != s1.obj;
 }
 
+inline bool operator<(const Selection& s0, const Selection& s1)
+{
+    return s0.type < s1.type || s0.obj < s1.obj;
```
Comment (375gnu) [inline src/celengine/selection.h]: I'm not sure that you do what you actually want.
1. I suppose you should compare objects only if `s0.type == s1.type`
1. obj is a pointer, so pointer comparison is nonsense, do you meant `*s0.obj < *s1.obj`?

File: src/celengine/category.cpp [hunk 31]
```diff
@@ -0,0 +1,114 @@
+#include <iostream>
+#include "category.h"
+
+UserCategory::UserCategory(const char *n, UserCategory *p) : m_name(n), m_parent(p) {}
+UserCategory::UserCategory(std::string n, UserCategory *p) : m_name(n), m_parent(p) {}
+UserCategory::~UserCategory() { cleanup(); }
```
[the remark this answers is not in the document]
Comment (375gnu) [reply src/celengine/category.cpp]: one more: if you use `std::string n`, then use `m_name(std::move(n))` to avoid copy creation

File: src/celengine/selection.h (hunk 30 above)
Comment (375gnu) [reply src/celengine/selection.h]: Pointer comparison must not be used here. It's valid for `!=` to ensure that s0 and s1 doesn't point to the same object. But not in this case.

```
if (s0.type != s1.type)
    return s0.type < s1.type;

// here goes more complicated code to work with s0.type == s1.type
// not an actual code, but pseudo code
switch (s0.type)
{
case Type_Nil: return false;
case Type_Body: return *static_cast<Body*>s0.obj < *static_cast<Body*>s1.obj;
...
}
```

Comment (pirogronian) [reply src/celengine/selection.h]: Hmm, I see no problem. Pointers comparisions is standard procedure in c++ , as pointer is just unsigned integer. It is fast and reliable. I see also no ease workaround. (Casting to typed pointer changes nothing.) I used it mamy times with many kinds of containers with any problem.

Comment (pirogronian) [reply src/celengine/selection.h]: Hmm, I see no problem. Pointers comparisions is standard procedure in c++ , as pointer is just unsigned integer. It is fast and reliable. I see also no ease workaround. (Casting to typed pointer changes nothing.) I used it mamy times with many kinds of containers with any problem.

Comment (375gnu) [reply src/celengine/selection.h]: Okay, let me explain. You have two planets, Earth and Saturn. Both are Type_Body, but Earth is located in memory at 0x0100 and Saturn is located at 0x0010. Your code says that Saturn < Earth. That's absurd.

Comment (pirogronian) [reply src/celengine/selection.h]: And this absurd is needed to store it in std::set. I hope nobody will try to compare two select objects, expecting anything else.

Comment (375gnu) [reply src/celengine/selection.h]: Hi @pirogronian, I've just realized that you won't need this operator if you use `unordered_set` instead of `set`. It seams that you actually don't need ordering.

Comment (pirogronian) [reply src/celengine/selection.h]: Segfaults eliminated, code simplified.

File: src/celengine/category.h [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+    const ObjectSet &objects() const { return m_objlist; }
+    const CategorySet &children() const { return m_catlist; }
+
+private:
+    bool _addObject(Selection);
+    bool _removeObject(Selection);
+    void _insertChild(UserCategory *);
+    bool removeChild(UserCategory *);
+    void setParent(UserCategory*);
+    static bool _deleteCategory(UserCategory *);
```
Comment (375gnu) [inline src/celengine/category.h]: extra space

File: src/celengine/category.h [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+    void cleanup();
+
+    static CategoryMap m_allcats;
+    static CategorySet m_roots;
+public:
+    static UserCategory *newCategory(const std::string&, UserCategory* parent = nullptr);
+    static UserCategory *find(const std::string&);
+
+    static const CategoryMap &getAll() { return m_allcats; }
+    static const CategorySet &getRoots() { return m_roots; }
```
Comment (375gnu) [inline src/celengine/category.h]: `getAll() const` and `getRoots() const`

File: src/celengine/category.h [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <set>
+#include <unordered_set>
+#include <map>
+#include "selection.h"
+
+class UserCategory
+{
+public:
+    struct Hasher
```
Comment (375gnu) [inline src/celengine/category.h]: i'm ok with the custom hasher, just a question - did it worked with a default one?

File: src/celengine/category.h (hunk 33 above)
Comment (pirogronian) [reply src/celengine/category.h]: `In file included from src/celengine/category.cpp:3:

src/celengine/category.h:59:40: error: static member function ‘static const CategoryMap& UserCategory::getAll()’ cannot have cv-qualifier

     static const CategoryMap &getAll() const { return m_allcats; }
                                        ^~~~~
src/celengine/category.h:60:42: error: static member function ‘static const CategorySet& UserCategory::getRoots()’ cannot have cv-qualifier

     static const CategorySet &getRoots() const { return m_roots; }
                                          ^~~~~`
Internet says, that static members may not be const (nor virtual nor volatile etc.)

File: src/celengine/category.h (hunk 34 above)
Comment (pirogronian) [reply src/celengine/category.h]: Default didn't exist. That's why I had to write own.

File: src/celengine/category.h (hunk 33 above)
Comment (375gnu) [reply src/celengine/category.h]: i missed `static`, sorry

**Comment by pirogronian** [comment]
1. Done.
2. The only information about dprintf I found here: [](http://man.freetechsecrets.com/dprintf.3.html). If this is it, I don't know how to use it.
3. Do you mean issues mentioned by you? I hope all are resolved. But I will do some tests yet.

**Comment by 375gnu** [comment]
>     2. The only information about dprintf I found here: http://man.freetechsecrets.com/dprintf.3.html. If this is it, I don't know how to use it.

I mean this macro: https://github.com/CelestiaProject/Celestia/blob/master/src/celutil/debug.h#L15

>     3. Do you mean issues mentioned by you? I hope all are resolved. But I will do some tests yet.

Incorrect style and **incorrect std::string usage in a constructor**
https://github.com/CelestiaProject/Celestia/pull/132/files#r237907787

Here you do extra operations. You added explicit return with value, but you don't use it.
https://github.com/CelestiaProject/Celestia/pull/132/files#r236565735

https://github.com/CelestiaProject/Celestia/pull/132/files#r237910377
https://github.com/CelestiaProject/Celestia/pull/132/files#r237910656

Optional. But proposed code is easier to understand.
https://github.com/CelestiaProject/Celestia/pull/132/files#r236568921

File: src/celestia/celx_internal.h (hunk 5 above)
Comment (pirogronian) [reply src/celestia/celx_internal.h]: I dont understand. This is as I write.

**Comment by pirogronian** [comment]
How to choose debug level?

**Comment by 375gnu** [comment]
> How to choose debug level?

compile with `-DDEBUG`. actuall https://github.com/CelestiaProject/Celestia/blob/master/src/celutil/debug.cpp is very simple

File: src/celestia/celx_internal.h (hunk 5 above)
Comment (375gnu) [reply src/celestia/celx_internal.h]: Ahhh, you want to return values of **your** push-es. okay.
